### ISSUE TITLE
Incremental Java compilation with Dagger failed with "Could not find class file"

### ISSUE THREAD

**Issue by hungvietnguyen** [google/dagger#3798]
### Steps to reproduce

These steps are adapted from: https://issuetracker.google.com/267743723#comment3.

1. `git clone https://github.com/kkoser/DaggerFactoryExample.git`
2. `git checkout 3788262846d043e06d477a2f2e11e305fd2e7bd8`
3. `./gradlew clean :app:compileDebugJavaWithJavac --no-build-cache`
4. `git checkout e2d5f1357b15aec7053288cb27289fba0576035c`
5. `./gradlew :app:compileDebugJavaWithJavac --no-build-cache` (Note: No `clean`)

The build will fail with
```
> Task :app:compileDebugJavaWithJavac FAILED
error: Could not find class file for 'com.example.mylibrary.MyBiggerDependency'.
DaggerFactoryExample/app/src/main/java/com/example/daggerfactoryexample/MainActivity.java:17: error: cannot find symbol
        DaggerActivityComponent.create().inject(this);
        ^
  symbol:   variable DaggerActivityComponent
  location: class MainActivity
2 errors
```

### Root cause
After step 3, the Dagger-generated file `DaggerFactoryExample/app/build/generated/ap_generated_sources/debug/out/com/example/daggerfactoryexample/DaggerActivityComponent.java` has references to both `com.example.mylibrary.MyBiggerDependency` and 
`com.example.mylibrary.MyDependency`. This means that Dagger is an **aggregating** annotation processor (one output file is generated from analyzing multiple input files).

However, `dagger-compiler` currently declares itself as an **isolating** annotation processor:
```
// dagger-compiler-2.45.jar!/META-INF/gradle/incremental.annotation.processors
dagger.internal.codegen.ComponentProcessor,ISOLATING
```

Declaring the wrong type will lead to incorrect incremental builds as reproduced above.

For more info on **aggregating** vs. **isolating** type, please see https://docs.gradle.org/current/userguide/java_plugin.html#sec:incremental_annotation_processing.

### Fix
Update Dagger from ISOLATING to AGGREGATING. (Also, it would be nice to have a regression test to confirm that it fixes the issue.)

**Comment by danysantiago** [comment]
I suspect something else might be going on but Dagger is behaving within the rules to be an `ISOLATING` processor since both `MyDependency` and `MyBiggerDependency` are reachable from the AST that triggered the processing (`ActivityComponent`). Which is fine for an isolating processor:

> "Isolating" processors have the following additional limitations:
> * They must make all decisions (code generation, validation messages) for an annotated type based on information reachable from its AST. This means you can analyze the types' super-class, method return types, annotations etc., even transitively. But you cannot make decisions based on unrelated elements in the RoundEnvironment.
> * They must provide exactly one originating element for each file generated with the Filer API. If zero or many originating elements are provided, Gradle will recompile all source files.

In this case the reachable AST is:
```
interface ActivityComponent
  method inject() with parameter of type MainActivity
    class MainActivity
      field myTransformer of type MyTransformer
        class MyTransformer
          constructor init() with parameter of type MyBiggerDependency
            class MyBiggerDependency
              constructor init() with parameter of type MyDependency
```
We have seen previous bugs in Gradle that break incremental annotation processing, for example https://github.com/gradle/gradle/issues/21203, so maybe something is going with incremental compilation / processing between the Kotlin library module and the Java module. I'll take a look at the sample app, but just wanted to mention that I don't think Dagger is wrongly classified as isolating.

**Comment by cogman** [comment]
Is there any work around here or do I just need to do a clean with every build?

**Comment by cogman** [comment]
Also, probably related to this issue

https://github.com/gradle/gradle/issues/24718

**Comment by kkoser** [comment]
Do we have resolution on where the issue lies here? I'm unclear if it's an AGP, Dagger, or Gradle issue, but would love to help get this resolved!
